primitives/src/lib.rs

```diff
@@ -1,19 +1,153 @@
 #![cfg_attr(not(feature = "std"), no_std, no_main)]
 
-pub use bounded_collections::{BoundedBTreeMap, BoundedBTreeSet, BoundedVec, ConstU32};
-// use scale::{Decode, Encode, MaxEncodedLen};
-
-pub mod cross_chain;
-pub mod storage_keys;
-
-// /// Some way of identifying an account on the chain.
-// #[derive(Encode, Decode, Debug, MaxEncodedLen)]
-// pub struct AccountId([u8; 32]);
-// Id used for identifying non-fungible collections.
-pub type CollectionId = u32;
-// Id used for identifying non-fungible items.
-pub type ItemId = u32;
-/// The maximum length of an attribute key.
-pub type KeyLimit = ConstU32<64>;
-/// The maximum approvals an item could have.
-pub type ApprovalsLimit = ConstU32<20>;
+use codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use scale_info::TypeInfo;
+pub use v0::*;
+
+/// The identifier of a token.
+pub type TokenId = u32;
+
+pub mod v0 {
+	pub use error::*;
+
+	use super::*;
+
+	mod error {
+		use super::*;
+
+		/// Reason why a Pop API call failed.
+		#[derive(Encode, Decode, Debug, Eq, PartialEq)]
+		#[cfg_attr(feature = "std", derive(TypeInfo))]
+		#[repr(u8)]
+		#[allow(clippy::unnecessary_cast)]
+		pub enum Error {
```

Comment:
Dumb question. In the chain extension, are we handling errors in the following way: 
- Infrastructure errors, such as BadOrigin, CannotLookup, etc. Basically, non module-errors are returned as a DispatchError instead of being wrapped into a status code. Why? Because these errors are for the developer, not for the contract.
- Use-case errors are returned via status code.

Is this correct? And if so, wouldn't that make most of this error enum unnecessary?

cc @evilrobot-01

Reply:
A contract calls into the runtime via the api. When this results into an error it is converted into a status code. Either from DispatchError -> status code or DispatchError -> Error -> status code. The first scenario happens for errors from pallet contracts and should always be in line with its respective Error variant. 

On the contract side the status code can be converted into Error or a use case error. The latter needs more thinking though as Tin and I are not sure of its usefulness or how the developers will use it ([ref](https://github.com/r0gue-io/pop-node/issues/99)). We might get a better view on this after the example Tin is working on, otherwise we see what the user feedback is or experience when we start to build with the api ourselves. What I'm trying to say is that we shouldn't put much focus on it. We will provide something which the developers can use and see whether it is useful besides having the Error.